### PR TITLE
Add VS Code Extension

### PR DESCRIPTION
**Pull request by Burhan-Q**
## Summary
Adds check to see if running in VS Code terminal. If VS Code terminal found, will additionally check if [Ultralytics-Snippets](https://marketplace.visualstudio.com/VSCode) extension is installed and if not, will show logging message. This was a recommendation from @glenn-jocher at the meeting on Monday 2024-08-05.

## Screenshots

Linux system without installed extension
image

Windows system with installed extension
image

Verified on Windows that uninstalling the extension would show prompt again.

## Notes
- I wasn't certain where the best location for the logging message would be, and open for suggestions to change
- Logging message shown at start of command or after first `ultralytics` import
- Checking for install looks for files, but these could be present even when user uninstalls extension
  - parsing extensions JSON file to check for install probably too much additional

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced VS Code support with a new extension for improved development efficiency using Ultralytics.

### 📊 Key Changes
- 📜 Added documentation for **Ultralytics-Snippets** VS Code extension.
- 🆕 Introduced a new file for the **VS Code Integration**.
- 🚀 Updated settings to include a prompt to install the VS Code extension when detected.
- 🛠️ Modified existing settings to incorporate the new `vscode_msg` setting.
- ⚙️ Adjusted the initialization of settings to a new version (`0.0.5`).

### 🎯 Purpose & Impact
- **Ease of Use:** The VS Code extension provides pre-built code snippets, reducing the need to write repetitive code and helping with quick access to common functionalities.
- **Productivity Boost:** With intelligent code completion and predefined snippets, developers can write cleaner, error-free code faster, enhancing overall productivity.
- **Improved Learning Curve:** New users and experienced developers alike can benefit from examples and streamlined development workflows, making it easier to adopt and use the Ultralytics Python package.
- **Accessibility:** Adding prompts to install the extension ensures that developers who use VS Code get the recommendations needed to enhance their development environment.

### CODE REVIEW

**Comment by Burhan-Q** [comment]
Included check for `"VSCODE_NO_RECOMMEND"` to give users ability to opt-out of prompt without installing extension. Also checking `.obsolete` file as extension directory does not get removed on uninstall, so users will be prompted again after uninstall.

File: ultralytics/utils/__init__.py [hunk 1]
```diff
@@ -1080,3 +1081,14 @@ def url2file(url):
...
+
+if IS_VSCODE and not os.environ.get("VSCODE_NO_RECOMMEND", False):
+    ext_path = (USER_CONFIG_DIR.parents[2] if WINDOWS else USER_CONFIG_DIR.parents[1]) / ".vscode/extensions"
+    obs_file = ext_path / ".obsolete"
+    ext = "ultralytics.ultralytics-snippets"
+    installed = ext_path.glob(f"{ext}*") and ext not in (obs_file.read_text("utf-8") if obs_file.exists() else "None")
+    if not installed:
+        LOGGER.info(
+            colorstr("VS Code terminal detected.\n")
+            + "  Enhance your Ultralytics experience by installing Ultralytics-Snippets for VS Code ⚡.\n"
```
Comment (ambitious-octopus) [inline ultralytics/utils/__init__.py:1093]: It would be helpful to add a link here!

**Review by ambitious-octopus — COMMENTED** [review]
Tested with vscode snap, flatpack and deb package. Works in all three cases!

**Comment by glenn-jocher** [comment]
@Burhan-Q ok nice. Let's drop all this into a function, and we need to think about how not to show this every time to keep from annoying users. 

Maybe we could show this only on the commands that return significant console output and that the user is likely to see, i.e. train/predict/export/val end commands that point to the docs pages. We also need to provide them a way to disable and notify them about it, likely in the settings, i.e. `yolo settings vscode_msg False`.

Otherwise I can see users getting extremely frustrated seeing this on each ultralytics import.

**Comment by Burhan-Q** [comment]
## TODOs
- [ ] Migrate code into a function
- [ ] Add a permanent settings key to enable/disable prompt, i.e. `yolo settings vscode False` to disable integration message, just like we disable the other integrations
 - [ ] Only show occasionally. 
   - Docs prompts are a good example, at the end of each mode command (i.e. at the end of the `train/predict/val` console output)
- [ ] Create a docs integration page that clearly explains how to install, uninstall, disable, enable etc. the integration, and link to that docs page directly (just like the existing message you see on `train/predict/val` etc.).

**Comment by Burhan-Q** [comment]
Now prompts after `predict` if not opted out, extension not installed, and VS Code detected.

image

**Comment by Burhan-Q** [comment]
## Images for docs integration page

<details><summary>partial snippet alias</summary>
<p>

image

</p>
</details> 

<details><summary>predict kwargs preview</summary>
<p>

image

</p>
</details> 

<details><summary>results loop preview</summary>
<p>

image

</p>
</details> 

<details><summary>fields updates</summary>
<p>

ultralytics-snippets_multi-update

</p>
</details>

**Comment by Burhan-Q** [comment]
I also updated `quickstart.md` table to include the new setting. While I was there, I added some links to the various integration pages, but left any existing external URLs. I also organized the links for the table at the bottom of the document for a cleaner doc. 

Other requested changes/features added. Will wait to mark as "ready" until I get some others to do a quick proof read of the new integration page.

**Comment by Burhan-Q** [comment]
@lakshanthad and @ambitious-octopus can I get a proof read on the documentation when you have some time please 🙌

**Review by ambitious-octopus — COMMENTED** [review]
The documentation looks perfect! But there is one feature that is not described: the extension is invoked even without writing ultra. 
As in these examples

image

image

image


Is it worth mentioning what do you think @Burhan-Q ?

**Comment by Burhan-Q** [comment]
@ambitious-octopus I did mention this a bit. The section on snippet completion.

> It's not required to type the full prefix of the snippet, or even from the start of the snippet. The snippets are named in the most descriptive way possible, but this means there could be a lot to type and that would be counterproductive if the aim is to move _faster_. Luckily VS Code lets users type `ultra.example-yolo-predict`, `example-yolo-predict`, `yolo-predict`, or even `ex-yolo-p` and still reach the intended snippet option!

File: docs/en/integrations/vscode-snippets.md [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+✅ Are you always forgetting the arguments or default values for the [export], [predict], [train], [track], or [val] methods?
+
+✅ Looking to get started with Ultralytics and wish you had an _easier_ way to reference or run code examples?
+
+✅ Want to speed up your development cycle when working with Ultralytics?
+
+If you use Visual Studio Code and answered 'yes' to any of the above, then the Ultralytics-snippets extension for VS Code is here to help! Read on to learn more about the extension, how to install it, and how to use it.
+
+## Inspired by the Ultralytics Community
```
[the remark this answers is not in the document]
Comment (Burhan-Q) [reply docs/en/integrations/vscode-snippets.md:30]: There is a link to the repo in a couple later sections. I think it's better to place it further down the page.

File: docs/en/integrations/vscode-snippets.md [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+comments: true
+description: An overview of how the Ultralytics-Snippets extension for Visual Studio Code can help developers accelerate their work with the Ultralytics Python package.
+keywords: Visual Studio Code, VS Code, deep learning, convolutional neural networks, computer vision, Python, code snippets, Ultralytics, developer productivity, machine learning, YOLO, developers, productivity, efficiency, learning, programming, IDE, code editor, developer utilities, programming tools
+---
+
+# Accelerate Your Computer Vision Projects with Ultralytics Code Snippets
+
+<p align="center">
+  <br>
+    <img src="https://github.com/Burhan-Q/ultralytics-snippets/assets/62214284/42ad0b17-e752-479c-9c6c-e451fffbe8b3" alt="Snippet Prediction Preview">
```
[the remark this answers is not in the document]
Comment (Burhan-Q) [reply docs/en/integrations/vscode-snippets.md:11]: I see what you're saying. Here's what my thoughts were when making this; I made an assumption that many new users would be viewing:
- Opening banner is supposed to provide context about what is going to happen.
- End banner is supposed help viewers understand that the operation is completed.
- Timer is there to show how fast running the example is when inserting via the snippet.

I won't argue that they draw attention, but this _was_ intentional. If there's a lot going on in the anmimation (that's relevant), it means that viewers will watch the loop multiple times to catch everything. 

@ambitious-octopus what do you mean abou the GIF being _slow_? As in a low frame rate or that it's too long (duration)? I tried to reduce the file size as much as possible without losing _too_ much quality or information/frames.

**Comment by lakshanthad** [comment]
> @lakshanthad and @ambitious-octopus can I get a proof read on the documentation when you have some time please 🙌

Yes. Will do this today and share feedback @Burhan-Q

**Comment by lakshanthad** [comment]
@Burhan-Q, I have reviewed the documentation and also tried the extension. It is already well documented along with great functionality of the snippet! Excellent work overall 🔥

I just have one small feedback about the GIF in the beginning. In the GIF, you are using `ultra.example-yolo`. How about we change to `yolo-p` since it will make it look easier to type since it is also supported?

**Comment by Burhan-Q** [comment]
@lakshanthad thanks man!

> I just have one small feedback about the GIF in the beginning. In the GIF, you are using ultra.example-yolo. How about we change to yolo-p since it will make it look easier to type?

It's possible to type only partial snippets. I tried to make the names _as obvious_ as possible, but I know that tends to make them long to type out. I mentioned in this comment https://github.com/ultralytics/ultralytics/pull/15027#issuecomment-2277505947 (with a quote from the doc) that the entire snippet prefix does not need to be used. 

Since it was brought up twice, I think that means I need to draw more attention to this. I think I'll include a "tip" admonition to make it stand out more.

**Comment by lakshanthad** [comment]
@Burhan-Q I know it is possible to type partial snippets which I saw in your documentation. What I was pointing out was to use 'yolo-p' in the GIF to make the GIF look even simple!

**Comment by Burhan-Q** [comment]
> @Burhan-Q I know it is possible to type partial snippets which I saw in your documentation. What I was pointing out was to use `yolo-p` in the GIF to make the GIF look even simple!

@lakshanthad I see what you mean. 🤔 I see what you're saying. I think I could do an A | B test for that in the future. For now, my preference is to keep the fully written (even tho I don't actually type all of it) prefix, so it's a bit more obvious. The input is appreciated!

**Comment by Burhan-Q** [comment]
image

**Pull request merged**